crates/sp-domains/src/lib.rs

```diff
@@ -479,9 +481,53 @@ pub enum RuntimeType {
     Evm,
 }
 
+impl PassBy for RuntimeType {
+    type PassBy = pass_by::Codec<Self>;
+}
+
 /// Type representing the runtime ID.
 pub type RuntimeId = u32;
 
+#[cfg(feature = "std")]
+pub trait GenerateGenesisStateRoot: Send + Sync {
+    /// Returns the state root of genesis block built from the runtime genesis config on success.
+    fn generate_genesis_state_root(
+        &self,
+        runtime_type: RuntimeType,
+        raw_runtime_genesis_config: Vec<u8>,
+    ) -> Option<H256>;
+}
+
+#[cfg(feature = "std")]
+sp_externalities::decl_extension! {
+    /// A domain genesis receipt extension.
+    pub struct GenesisReceiptExtension(std::sync::Arc<dyn GenerateGenesisStateRoot>);
```

Comment:
Confused here, `GenesisReceiptExtension` should match with `GenerateGenesisReceipt` and `generate_genesis_receipt`, the genesis receipt is what expected from the runtime `instantiate_domain` call instead of just the genesis state root.

Reply:
The reason for returning the genesis state root instead of the full receipt is that the generic parameter is unsupported in the runtime interface, you can't return the typed receipt directly, so I decide to only calculate the state root and let the runtime assemble the final receipt.